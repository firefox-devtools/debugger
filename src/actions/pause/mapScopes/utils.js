// @flow

import type { Location } from "../../../types";

export function locColumn(loc: Location): number {
  if (typeof loc.column !== "number") {
    // This shouldn't really happen with locations from the AST, but
    // the datatype we are using allows null/undefined column.
    return 0;
  }

  return loc.column;
}
