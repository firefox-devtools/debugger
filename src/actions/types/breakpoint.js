// @flow

import type { Breakpoint, Location } from "../../types";

import type { PromiseAction } from "../utils/middleware/promise";

type AddBreakpointResult = {
  previousLocation: Location,
  breakpoint: Breakpoint
};

export type BreakpointAction =
  | PromiseAction<
      {|
        +type: "ADD_BREAKPOINT",
        +breakpoint: Breakpoint,
        +condition?: string
      |},
      AddBreakpointResult
    >
  | PromiseAction<{|
      +type: "REMOVE_BREAKPOINT",
      +breakpoint: Breakpoint,
      +disabled: boolean
    |}>
  // for simulating a successful server request
  | {|
      +type: "REMOVE_BREAKPOINT",
      +breakpoint: Breakpoint,
      +status: "done"
    |}
  | {|
      +type: "SET_BREAKPOINT_CONDITION",
      +breakpoint: Breakpoint
    |}
  | PromiseAction<{|
      +type: "TOGGLE_BREAKPOINTS",
      +shouldDisableBreakpoints: boolean
    |}>
  | {|
      +type: "SYNC_BREAKPOINT",
      +breakpoint: ?Breakpoint,
      +previousLocation: Location
    |}
  | PromiseAction<
      {|
        +type: "ENABLE_BREAKPOINT",
        +breakpoint: Breakpoint
      |},
      AddBreakpointResult
    >
  | {|
      +type: "DISABLE_BREAKPOINT",
      +breakpoint: Breakpoint
    |}
  | {|
      +type: "DISABLE_ALL_BREAKPOINTS",
      +breakpoints: Breakpoint[]
    |}
  | {|
      +type: "ENABLE_ALL_BREAKPOINTS",
      +breakpoints: Breakpoint[]
    |}
  | {|
      +type: "REMAP_BREAKPOINTS",
      +breakpoints: Breakpoint[]
    |};
