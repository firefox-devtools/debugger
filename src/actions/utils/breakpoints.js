import { getBreakpoint } from "../../selectors";

export async function formatClientBreakpoint(
  clientBreakpoint,
  sourceMaps,
  location
) {
  const clientOriginalLocation = await sourceMaps.getOriginalLocation(
    clientBreakpoint.actualLocation
  );

  // make sure that we are re-adding the same type of breakpoint. Column
  // or line
  const actualLocation = clientOriginalLocation;

  // the generatedLocation might have slid, so now we can adjust it
  const generatedLocation = clientBreakpoint.actualLocation;

  const { id, hitCount } = clientBreakpoint;
  return { id, actualLocation, hitCount, generatedLocation };
}

export function breakpointExists(state, location: Location) {
  const currentBp = getBreakpoint(state, location);
  return currentBp && !currentBp.disabled;
}

export function createBreakpoint(location: Object, overrides: Object = {}) {
  const { condition, disabled, generatedLocation } = overrides;
  const properties = {
    condition: condition || null,
    disabled: disabled || false,
    generatedLocation,
    location
  };

  return properties;
}

export async function getGeneratedLocation(source, sourceMaps, location) {
  if (!sourceMaps.isOriginalId(location.sourceId)) {
    return location;
  }

  return await sourceMaps.getGeneratedLocation(location, source.toJS());
}

export function equalizeLocationColumn(location, referenceLocation) {
  if (referenceLocation.column) {
    return location;
  }
  return { ...location, column: undefined };
}
