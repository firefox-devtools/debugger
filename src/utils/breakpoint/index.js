/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import { sortBy } from "lodash";

import { getBreakpoint } from "../../selectors";
import assert from "../assert";
import { features } from "../prefs";

export { getASTLocation, findScopeByName } from "./astBreakpointLocation";

import type { FormattedBreakpoint } from "../../selectors/breakpointSources";
import type {
  SourceLocation,
  PendingLocation,
  Breakpoint,
  PendingBreakpoint
} from "../../types";

import type { State } from "../../reducers/types";

// Return the first argument that is a string, or null if nothing is a
// string.
export function firstString(...args: string[]) {
  for (const arg of args) {
    if (typeof arg === "string") {
      return arg;
    }
  }
  return null;
}

export function locationMoved(
  location: SourceLocation,
  newLocation: SourceLocation
) {
  return (
    location.line !== newLocation.line || location.column !== newLocation.column
  );
}

export function makeLocationId(location: SourceLocation) {
  const { sourceId, line, column } = location;
  const columnString = column || "";
  return `${sourceId}:${line}:${columnString}`;
}

export function getLocationWithoutColumn(location: SourceLocation) {
  const { sourceId, line } = location;
  return `${sourceId}:${line}`;
}

export function makePendingLocationId(location: SourceLocation) {
  assertPendingLocation(location);
  const { sourceUrl, line, column } = location;
  const sourceUrlString = sourceUrl || "";
  const columnString = column || "";

  return `${sourceUrlString}:${line}:${columnString}`;
}

export function assertBreakpoint(breakpoint: Breakpoint) {
  assertLocation(breakpoint.location);
  assertLocation(breakpoint.generatedLocation);
}

export function assertPendingBreakpoint(pendingBreakpoint: PendingBreakpoint) {
  assertPendingLocation(pendingBreakpoint.location);
  assertPendingLocation(pendingBreakpoint.generatedLocation);
}

export function assertLocation(location: SourceLocation) {
  assertPendingLocation(location);
  const { sourceId } = location;
  assert(!!sourceId, "location must have a source id");
}

export function assertPendingLocation(location: PendingLocation) {
  assert(!!location, "location must exist");

  const { sourceUrl } = location;

  // sourceUrl is null when the source does not have a url
  assert(sourceUrl !== undefined, "location must have a source url");
  assert(location.hasOwnProperty("line"), "location must have a line");
  assert(
    location.hasOwnProperty("column") != null,
    "location must have a column"
  );
}

// syncing
export function breakpointAtLocation(
  breakpoints: Breakpoint[],
  { line, column }: SourceLocation
) {
  return breakpoints.find(breakpoint => {
    const sameLine = breakpoint.location.line === line;
    if (!sameLine) {
      return false;
    }

    // NOTE: when column breakpoints are disabled we want to find
    // the first breakpoint
    if (!features.columnBreakpoints) {
      return true;
    }

    return breakpoint.location.column === column;
  });
}

export function breakpointExists(state: State, location: SourceLocation) {
  const currentBp = getBreakpoint(state, location);
  return currentBp && !currentBp.disabled;
}

export function createBreakpoint(
  location: SourceLocation,
  overrides: Object = {}
): Breakpoint {
  const {
    condition,
    disabled,
    hidden,
    generatedLocation,
    astLocation,
    id,
    text,
    originalText,
    log
  } = overrides;

  const defaultASTLocation = {
    name: undefined,
    offset: location,
    index: 0
  };
  const properties = {
    id,
    condition: condition || null,
    log: log || false,
    disabled: disabled || false,
    hidden: hidden || false,
    loading: false,
    astLocation: astLocation || defaultASTLocation,
    generatedLocation: generatedLocation || location,
    location,
    text,
    originalText
  };

  return properties;
}

export function createXHRBreakpoint(
  path: string,
  method: string,
  overrides?: Object = {}
) {
  const properties = {
    path,
    method,
    disabled: false,
    loading: false,
    text: `URL contains "${path}"`
  };

  return { ...properties, ...overrides };
}

function createPendingLocation(location: PendingLocation) {
  const { sourceUrl, line, column } = location;
  return { sourceUrl, line, column };
}

export function createPendingBreakpoint(bp: Breakpoint) {
  const pendingLocation = createPendingLocation(bp.location);
  const pendingGeneratedLocation = createPendingLocation(bp.generatedLocation);

  assertPendingLocation(pendingLocation);

  return {
    condition: bp.condition,
    log: bp.log,
    disabled: bp.disabled,
    location: pendingLocation,
    astLocation: bp.astLocation,
    generatedLocation: pendingGeneratedLocation
  };
}

export function sortFormattedBreakpoints(breakpoints: FormattedBreakpoint[]) {
  return _sortBreakpoints(breakpoints, "selectedLocation");
}

export function sortBreakpoints(breakpoints: Breakpoint[]) {
  return _sortBreakpoints(breakpoints, "location");
}

function _sortBreakpoints(
  breakpoints: Array<Object>,
  property: string
): Array<Object> {
  // prettier-ignore
  return sortBy(
    breakpoints,
    [
      // Priority: line number, undefined column, column number
      `${property}.line`,
      bp => {
        return bp[property].column === undefined || bp[property].column;
      }
    ]
  );
}
