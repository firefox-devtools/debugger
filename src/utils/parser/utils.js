// @flow

const babylon = require("babylon");
const traverse = require("babel-traverse").default;
const t = require("babel-types");
const { isDevelopment } = require("devtools-config");
const toPairs = require("lodash/toPairs");
const get = require("lodash/get");
const isEmpty = require("lodash/isEmpty");

import type { SourceText, Location, Frame, TokenResolution } from "../../types";

const ASTs = new Map();

const symbolDeclarations = new Map();

type ASTLocation = {
  start: {
    line: number,
    column: number
  },
  end: {
    line: number,
    column: number
  }
};

export type SymbolDeclaration = {
  name: string,
  location: ASTLocation
};

export type FormattedSymbolDeclaration = {
  id: string,
  title: string,
  subtitle: string,
  value: string,
  location: ASTLocation
};

export type SymbolDeclarations = {
  functions: Array<FormattedSymbolDeclaration>,
  variables: Array<FormattedSymbolDeclaration>
};

function _parse(code) {
  return babylon.parse(code, {
    sourceType: "module",

    plugins: ["jsx", "flow"]
  });
}

function parse(text: string) {
  let ast;
  if (!text) {
    return;
  }

  try {
    ast = _parse(text);
  } catch (error) {
    if (isDevelopment()) {
      console.warn("parse failed", text);
    }

    ast = {};
  }

  return ast;
}

function getAst(sourceText: SourceText) {
  if (ASTs.has(sourceText.id)) {
    return ASTs.get(sourceText.id);
  }

  const ast = parse(sourceText.text);
  ASTs.set(sourceText.id, ast);
  return ast;
}

function getFunctionName(path) {
  if (path.node.id) {
    return path.node.id.name;
  }

  const parent = path.parent;
  if (parent.type == "ObjectProperty") {
    return parent.key.name;
  }

  if (parent.type == "ObjectExpression" || path.node.type == "ClassMethod") {
    return path.node.key.name;
  }

  if (parent.type == "VariableDeclarator") {
    return parent.id.name;
  }

  if (parent.type == "AssignmentExpression") {
    if (parent.left.type == "MemberExpression") {
      return parent.left.property.name;
    }

    return parent.left.name;
  }

  return "anonymous";
}

function isFunction(path) {
  return t.isFunction(path) ||
    t.isArrowFunctionExpression(path) ||
    t.isObjectMethod(path) ||
    t.isClassMethod(path);
}

function formatSymbol(symbol: SymbolDeclaration): FormattedSymbolDeclaration {
  return {
    id: `${symbol.name}:${symbol.location.start.line}`,
    title: symbol.name,
    subtitle: `:${symbol.location.start.line}`,
    value: symbol.name,
    location: symbol.location
  };
}

function getVariableNames(path) {
  if (t.isObjectProperty(path) && !isFunction(path.node.value)) {
    return [
      formatSymbol({
        name: path.node.key.name,
        location: path.node.loc
      })
    ];
  }

  if (!path.node.declarations) {
    return path.node.params.map(dec =>
      formatSymbol({
        name: dec.name,
        location: dec.loc
      }));
  }

  return path.node.declarations.map(dec =>
    formatSymbol({
      name: dec.id.name,
      location: dec.loc
    }));
}

function isVariable(path) {
  return t.isVariableDeclaration(path) ||
    (isFunction(path) && path.node.params.length) ||
    (t.isObjectProperty(path) && !isFunction(path.node.value));
}

function getMemberExpression(root) {
  function _getMemberExpression(node, expr) {
    if (t.isMemberExpression(node)) {
      expr = [node.property.name].concat(expr);
      return _getMemberExpression(node.object, expr);
    }

    if (t.isThisExpression(node)) {
      return ["this"].concat(expr);
    }
    return [node.name].concat(expr);
  }

  return _getMemberExpression(root, []);
}

function getSymbols(source: SourceText): SymbolDeclarations {
  if (symbolDeclarations.has(source.id)) {
    const symbols = symbolDeclarations.get(source.id);
    if (symbols) {
      return symbols;
    }
  }

  const ast = getAst(source);
  const symbols = { functions: [], variables: [] };

  traverse(ast, {
    enter(path) {
      if (isVariable(path)) {
        symbols.variables.push(...getVariableNames(path));
      }

      if (isFunction(path)) {
        symbols.functions.push(
          formatSymbol({
            name: getFunctionName(path),
            location: path.node.loc
          })
        );
      }

      if (t.isClassDeclaration(path)) {
        symbols.variables.push(
          formatSymbol({
            name: path.node.id.name,
            location: path.node.loc
          })
        );
      }
    }
  });

  symbolDeclarations.set(source.id, symbols);
  return symbols;
}

function resolveExpression(path, token: string, location: Location): ?Object {
  const node = path.node;
  if (
    t.isMemberExpression(node) &&
    node.property.name === token &&
    nodeContainsLocation({ node, location })
  ) {
    const expr = getMemberExpression(node);
    return {
      value: expr.join("."),
      location: node.loc
    };
  }

  return null;
}

function resolveScope(path, location: Location) {
  const node = path.node;
  if (
    (isFunction(path) || t.isProgram(path)) &&
    nodeContainsLocation({ node, location })
  ) {
    return path;
  }
}

// Resolves a token (at location) in the source to determine if it is in scope
// of the given frame and the expression (if any) to which it belongs
function resolveToken(
  source: SourceText,
  token: string,
  location: Location,
  frame: Frame
): ?TokenResolution {
  const ast = getAst(source);
  const scopes = [];
  let expression = null;
  let inScope = false;

  if (isEmpty(ast)) {
    return;
  }

  traverse(ast, {
    enter(path) {
      let scope = null;

      // if we haven't found an expression yet, determine if the token is part
      // of one
      if (!expression) {
        expression = resolveExpression(path, token, location);
      }

      // determine if the current path is a function or program containing the
      // frame
      scope = resolveScope(path, frame.location);
      if (scope) {
        scopes.unshift(scope);
      }
    }
  });

  // determine if the narrowest scope contains the token's location
  inScope = nodeContainsLocation({ node: scopes[0].node, location });

  return {
    expression,
    inScope
  };
}

function nodeContainsLocation({ node, location }) {
  const { start, end } = node.loc;
  const { line, column } = location;

  return !(start.line > line ||
    (start.line === line && start.column > column) ||
    end.line < line ||
    (end.line === line && end.column < column));
}

function getPathClosestToLocation(source: SourceText, location: Location) {
  const ast = getAst(source);
  let pathClosestToLocation = null;

  traverse(ast, {
    enter(path) {
      if (nodeContainsLocation({ node: path.node, location })) {
        pathClosestToLocation = path;
      }
    }
  });

  return pathClosestToLocation;
}

function getVariablesInScope(source: SourceText, location: Location) {
  const path = getPathClosestToLocation(source, location);
  const bindings = get(path, "scope.bindings", {});

  return toPairs(bindings).map(([name, binding]) => ({
    name,
    references: binding.referencePaths
  }));
}

module.exports = {
  getSymbols,
  getPathClosestToLocation,
  getVariablesInScope,
  resolveToken
};
