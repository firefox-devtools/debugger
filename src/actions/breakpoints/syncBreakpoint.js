import {
  locationMoved,
  createBreakpoint,
  assertBreakpoint,
  assertPendingBreakpoint,
  findScopeByName
} from "../../utils/breakpoint";

import { getGeneratedLocation } from "../../utils/source-maps";
import { originalToGeneratedId } from "devtools-source-map";
import { getSource } from "../../selectors";

async function makeScopedLocation({ name, offset }, location, source) {
  const scope = await findScopeByName(source, name);
  // fallback onto the location line, if the scope is not found
  // note: we may at some point want to delete the breakpoint if the scope
  // disappears
  const line = scope ? scope.location.start.line + offset.line : location.line;
  return {
    line,
    column: location.column,
    sourceUrl: source.url,
    sourceId: source.id
  };
}

function createSyncData(
  pendingBreakpoint,
  location,
  generatedLocation,
  previousLocation = null
) {
  const overrides = { ...pendingBreakpoint, generatedLocation };
  const breakpoint = createBreakpoint(location, overrides);

  assertBreakpoint(breakpoint);
  return { breakpoint, previousLocation };
}

// we have three forms of syncing: disabled syncing, existing server syncing
// and adding a new breakpoint
export async function syncClientBreakpoint(
  getState,
  client,
  sourceMaps,
  sourceId,
  pendingBreakpoint: PendingBreakpoint
) {
  assertPendingBreakpoint(pendingBreakpoint);

  const source = getSource(getState(), sourceId).toJS();
  const generatedSourceId = sourceMaps.isOriginalId(sourceId)
    ? originalToGeneratedId(sourceId)
    : sourceId;

  const { location, astLocation } = pendingBreakpoint;
  const previousLocation = { ...location, sourceId };

  const scopedLocation = await makeScopedLocation(
    astLocation,
    previousLocation,
    source
  );

  const scopedGeneratedLocation = await getGeneratedLocation(
    getState(),
    source,
    scopedLocation,
    sourceMaps
  );

  // this is the generatedLocation of the pending breakpoint, with
  // the source id updated to reflect the new connection
  const generatedLocation = {
    ...pendingBreakpoint.generatedLocation,
    sourceId: generatedSourceId
  };

  const isSameLocation = !locationMoved(
    generatedLocation,
    scopedGeneratedLocation
  );

  const existingClient = client.getBreakpointByLocation(generatedLocation);

  /** ******* CASE 1: No server change ***********/
  // early return if breakpoint is disabled or we are in the sameLocation
  // send update only to redux
  if (pendingBreakpoint.disabled || (existingClient && isSameLocation)) {
    return createSyncData(
      pendingBreakpoint,
      scopedLocation,
      scopedGeneratedLocation
    );
  }

  // clear server breakpoints if they exist and we have moved
  if (existingClient) {
    await client.removeBreakpoint(generatedLocation);
  }

  /** ******* Case 2: Add New Breakpoint ***********/
  // If we are not disabled, set the breakpoint on the server and get
  // that info so we can set it on our breakpoints.
  const clientBreakpoint = await client.setBreakpoint(
    scopedGeneratedLocation,
    pendingBreakpoint.condition,
    sourceMaps.isOriginalId(sourceId)
  );

  // the breakpoint might have slid server side, so we want to get the location
  // based on the server's return value
  const newGeneratedLocation = clientBreakpoint.actualLocation;
  const newLocation = await sourceMaps.getOriginalLocation(
    newGeneratedLocation
  );

  return createSyncData(
    pendingBreakpoint,
    newLocation,
    newGeneratedLocation,
    previousLocation
  );
}
