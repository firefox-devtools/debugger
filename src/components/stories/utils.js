const renderComponentFromFixture =
      require("../../test/utils/renderComponentFromFixture");

const { storiesOf } = require("@kadira/storybook");

module.exports = {
  renderComponent: renderComponentFromFixture,
  storiesOf
};
