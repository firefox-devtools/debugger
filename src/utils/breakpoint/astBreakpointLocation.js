// @flow

import { getSymbols } from "../parser";
import { containsPosition } from "../parser/utils/contains";

import type { Scope } from "../parser/types";
import type { Location, Source } from "debugger-html";

export function findClosestScope(functions: Scope[], location: Location) {
  return functions.reduce((found, currNode) => {
    if (
      currNode.name === "anonymous" ||
      !containsPosition(currNode.location, location)
    ) {
      return found;
    }

    if (!found) {
      return currNode;
    }

    if (found.location.start.line > currNode.location.start.line) {
      return found;
    }
    if (
      found.location.start.line === currNode.location.start.line &&
      found.location.start.column > currNode.location.start.column
    ) {
      return found;
    }

    return currNode;
  }, null);
}

export async function getASTLocation(source: Source, location: Location) {
  const symbols = await getSymbols(source);
  const functions = [...symbols.functions];

  const scope = findClosestScope(functions, location);
  if (scope) {
    // we only record the line, but at some point we may
    // also do column offsets
    const line = location.line - scope.location.start.line;
    return {
      name: scope.name,
      offset: { line }
    };
  }
  return { name: undefined, offset: location };
}

export async function findScopeByName(source: Source, name: String) {
  const symbols = await getSymbols(source);
  const functions = symbols.functions;

  return functions.find(node => node.name === name);
}
