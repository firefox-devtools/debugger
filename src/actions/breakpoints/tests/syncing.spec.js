/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

jest.mock("../../../utils/source-maps", () => ({
  getGeneratedLocation: jest.fn()
}));
import { getGeneratedLocation } from "../../../utils/source-maps";

jest.mock("../../../utils/prefs", () => ({
  prefs: {
    expressions: [],
    pendingBreakpoints: {}
  },
  features: {
    replay: false
  },
  clear: jest.fn()
}));

import {
  createStore,
  selectors,
  actions,
  makeSource
} from "../../../utils/test-head";

import { makeBreakpointId } from "../../../utils/breakpoint";

jest.mock("../../../utils/breakpoint/astBreakpointLocation", () => ({
  findScopeByName: jest.fn(),
  getASTLocation: jest.fn()
}));

// eslint-disable-next-line
import { findScopeByName } from "../../../utils/breakpoint/astBreakpointLocation";

import { syncBreakpointPromise } from "../../breakpoints/syncBreakpoint.js";

function setBreakpoint(location, condition) {
  const actualLocation = { ...location, line: location.line };

  return Promise.resolve({
    id: makeBreakpointId(location),
    actualLocation,
    condition
  });
}

const clientBreakpoint = {
  id: "foo",
  actualLocation: {
    sourceUrl: "http://localhost:8000/gen.js",
    sourceId: "gen.js",
    line: 3,
    column: undefined
  }
};

const threadClient = {
  getBreakpointByLocation: () => clientBreakpoint,
  setBreakpoint,
  removeBreakpoint: jest.fn()
};

const sourceMaps = {
  getOriginalLocation: () => ({
    sourceId: "magic.js",
    sourceUrl: "http://localhost:8000/magic.js",
    line: 3,
    column: undefined
  }),
  getOriginalURLs: () => {},
  isOriginalId: () => true,
  getGeneratedLocation: () => ({})
};

function pendingBreakpoint(overrides) {
  return {
    location: {
      sourceId: "magic.js",
      sourceUrl: "http://localhost:8000/magic.js",
      line: 3,
      column: undefined
    },
    generatedLocation: {
      sourceId: "gen.js",
      sourceUrl: "http://localhost:8000/gen.js",
      line: 3,
      column: undefined
    },
    astLocation: {
      name: undefined,
      offset: {
        line: 3
      },
      index: 0
    },
    options: {
      logValue: "",
      hidden: false
    },
    disabled: false,
    loading: false,
    text: "",
    ...overrides
  };
}

function newGeneratedLocation(line) {
  return {
    sourceUrl: "http://localhost:8000/gen.js",
    sourceId: "gen.js",
    line,
    column: undefined
  };
}

describe("loading the debugger", () => {
  it("loads the initial breakpoint state", async () => {
    (getGeneratedLocation: any).mockImplementation(() =>
      newGeneratedLocation(3)
    );

    const { dispatch, getState } = createStore(threadClient, {}, sourceMaps);

    // add a source without the breakpoints
    const reloadedSource = makeSource("magic.js");
    await dispatch(actions.newSource(reloadedSource));

    expect(selectors.getBreakpointCount(getState())).toEqual(0);
    // manually sync
    const update = await syncBreakpointPromise(
      getState,
      threadClient,
      sourceMaps,
      dispatch,
      reloadedSource.source.id,
      pendingBreakpoint()
    );

    expect(threadClient.removeBreakpoint.mock.calls).toHaveLength(0);
    expect(update).toMatchSnapshot();
  });

  it("loads the initial breakpoint state with a changed file", async () => {
    const location = { line: 9, column: 0 };
    const generated = 3;
    (getGeneratedLocation: any).mockImplementation(() =>
      newGeneratedLocation(generated)
    );
    (findScopeByName: any).mockImplementation(() => ({
      location: { start: location }
    }));

    const { dispatch, getState } = createStore(threadClient, {}, sourceMaps);

    // add a source without the breakpoints
    const reloadedSource = makeSource("magic.js");
    await dispatch(actions.newSource(reloadedSource));

    expect(selectors.getBreakpointCount(getState())).toEqual(0);
    // manually sync
    const update = await syncBreakpointPromise(
      getState,
      threadClient,
      sourceMaps,
      dispatch,
      reloadedSource.source.id,
      pendingBreakpoint()
    );

    expect(threadClient.removeBreakpoint.mock.calls).toHaveLength(0);
    expect(update && update.breakpoint && update.breakpoint.location.line).toBe(
      location.line + generated
    );
    expect(update).toMatchSnapshot();
  });
});

describe("reloading debuggee", () => {
  beforeEach(() => {
    const location = { line: 0, column: 0 };
    (getGeneratedLocation: any).mockImplementation(() =>
      newGeneratedLocation(3)
    );
    (findScopeByName: any).mockImplementation(() => ({
      location: { start: location }
    }));
  });

  it("syncs with unchanged source with an existing BP", async () => {
    const { dispatch, getState } = createStore(threadClient, {}, sourceMaps);

    // add a source without the breakpoints
    const reloadedSource = makeSource("magic.js");
    const loc1 = {
      sourceId: "magic.js",
      sourceUrl: "http://localhost:8000/magic.js",
      line: 3,
      column: undefined
    };
    const generatedSource = makeSource("gen.js");
    await dispatch(actions.newSource(reloadedSource));
    await dispatch(actions.newSource(generatedSource));
    await dispatch(actions.addBreakpoint(loc1));

    // manually sync
    const update = await syncBreakpointPromise(
      getState,
      threadClient,
      sourceMaps,
      dispatch,
      reloadedSource.source.id,
      pendingBreakpoint({ location: loc1 })
    );
    expect(threadClient.removeBreakpoint.mock.calls).toHaveLength(0);
    expect(update).toMatchSnapshot();
  });

  it("updates a corresponding breakpoint for a changed source", async () => {
    /*
      This is a complicated test, so bear with us. In this scenario,

      1. the user has a small app with a generated file gen.js
      and original file magic.js.
      2. The user adds a breakpoint in magic.js#3, which maps to gen.js#3
      3. The user edits their code and reloads
      4. when magic.js is added, the debugger syncs the saved breakpoint
         4.a. the debugger checks to see if gen.js#3 still points to magic.js#3,
              it now points to gen.js#1 so it removes the old
              breakpoint and creates a new one
    */

    // here we are mocking out what happens when the source changed, and the
    // new line for originalSource line 3, is the generated Source line 5

    (getGeneratedLocation: any).mockImplementation(() =>
      newGeneratedLocation(5)
    );
    // end mocking out

    const { dispatch, getState } = createStore(threadClient, {}, sourceMaps);

    // add a source without the breakpoints
    const reloadedSource = makeSource("magic.js");
    await dispatch(actions.newSource(reloadedSource));

    const generatedSource = makeSource("gen.js");
    await dispatch(actions.newSource(generatedSource));

    // manually sync
    const update = await syncBreakpointPromise(
      getState,
      threadClient,
      sourceMaps,
      dispatch,
      reloadedSource.source.id,
      pendingBreakpoint()
    );
    expect(threadClient.removeBreakpoint.mock.calls).toHaveLength(1);
    expect(findScopeByName).toHaveBeenCalled();
    expect(update).toMatchSnapshot();
  });

  it("syncs with changed source and an existing disabled BP", async () => {
    (getGeneratedLocation: any).mockImplementationOnce(() =>
      newGeneratedLocation(5)
    );

    const { dispatch, getState } = createStore(threadClient, {}, sourceMaps);

    const reloadedSource = makeSource("magic.js");
    await dispatch(actions.newSource(reloadedSource));

    const generatedSource = makeSource("gen.js");
    await dispatch(actions.newSource(generatedSource));

    const location = {
      sourceId: reloadedSource.source.id,
      line: 3,
      column: undefined
    };

    const breakpoint = await dispatch(actions.addBreakpoint(location));
    await dispatch(actions.disableBreakpoint(breakpoint));

    (getGeneratedLocation: any).mockImplementationOnce(() =>
      newGeneratedLocation(1)
    );

    await dispatch(
      actions.syncBreakpoint(
        reloadedSource.source.id,
        pendingBreakpoint({ disabled: true })
      )
    );

    expect(selectors.getPendingBreakpoints(getState())).toMatchSnapshot();
  });
});
