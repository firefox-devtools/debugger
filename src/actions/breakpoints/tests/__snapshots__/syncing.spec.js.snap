// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading the debugger loads the initial breakpoint state 1`] = `
Object {
  "breakpoint": Object {
    "astLocation": Object {
      "name": undefined,
      "offset": Object {
        "line": 3,
      },
    },
    "condition": null,
    "disabled": false,
    "generatedLocation": Object {
      "column": undefined,
      "line": 3,
      "sourceId": "gen.js",
      "sourceUrl": "http://localhost:8000/gen.js",
    },
    "hidden": false,
    "id": "foo",
    "loading": false,
    "location": Object {
      "column": undefined,
      "line": 3,
      "sourceId": "magic.js",
      "sourceUrl": "http://localhost:8000/examples/magic.js",
    },
    "log": false,
    "originalText": "",
    "text": "",
  },
  "previousLocation": Object {
    "column": undefined,
    "line": 3,
    "sourceId": "magic.js",
    "sourceUrl": "http://localhost:8000/magic.js",
  },
}
`;

exports[`loading the debugger loads the initial breakpoint state with a changed file 1`] = `
Object {
  "breakpoint": Object {
    "astLocation": Object {
      "name": undefined,
      "offset": Object {
        "line": 3,
      },
    },
    "condition": null,
    "disabled": false,
    "generatedLocation": Object {
      "column": undefined,
      "line": 3,
      "sourceId": "gen.js",
      "sourceUrl": "http://localhost:8000/gen.js",
    },
    "hidden": false,
    "id": "foo",
    "loading": false,
    "location": Object {
      "column": undefined,
      "line": 12,
      "sourceId": "magic.js",
      "sourceUrl": "http://localhost:8000/examples/magic.js",
    },
    "log": false,
    "originalText": "",
    "text": "",
  },
  "previousLocation": Object {
    "column": undefined,
    "line": 3,
    "sourceId": "magic.js",
    "sourceUrl": "http://localhost:8000/magic.js",
  },
}
`;

exports[`reloading debuggee syncs with changed source and an existing disabled BP 1`] = `
Object {
  "http://localhost:8000/examples/magic.js:3:": Object {
    "astLocation": Object {
      "name": undefined,
      "offset": Object {
        "line": 3,
      },
    },
    "condition": null,
    "disabled": true,
    "generatedLocation": Object {
      "column": undefined,
      "line": 1,
      "sourceUrl": "http://localhost:8000/gen.js",
    },
    "location": Object {
      "column": undefined,
      "line": 3,
      "sourceUrl": "http://localhost:8000/examples/magic.js",
    },
    "log": false,
  },
}
`;

exports[`reloading debuggee syncs with unchanged source with an existing BP 1`] = `
Object {
  "breakpoint": Object {
    "astLocation": Object {
      "name": undefined,
      "offset": Object {
        "line": 3,
      },
    },
    "condition": null,
    "disabled": false,
    "generatedLocation": Object {
      "column": undefined,
      "line": 3,
      "sourceId": "gen.js",
      "sourceUrl": "http://localhost:8000/gen.js",
    },
    "hidden": false,
    "id": "foo",
    "loading": false,
    "location": Object {
      "column": undefined,
      "line": 3,
      "sourceId": "magic.js",
      "sourceUrl": "http://localhost:8000/examples/magic.js",
    },
    "log": false,
    "originalText": "",
    "text": "",
  },
  "previousLocation": Object {
    "column": undefined,
    "line": 3,
    "sourceId": "magic.js",
    "sourceUrl": "http://localhost:8000/magic.js",
  },
}
`;

exports[`reloading debuggee updates a corresponding breakpoint for a changed source 1`] = `
Object {
  "breakpoint": Object {
    "astLocation": Object {
      "name": undefined,
      "offset": Object {
        "line": 3,
      },
    },
    "condition": null,
    "disabled": false,
    "generatedLocation": Object {
      "column": undefined,
      "line": 5,
      "sourceId": "gen.js",
      "sourceUrl": "http://localhost:8000/gen.js",
    },
    "hidden": false,
    "id": "gen.js:5:",
    "loading": false,
    "location": Object {
      "column": undefined,
      "line": 3,
      "sourceId": "magic.js",
      "sourceUrl": "http://localhost:8000/magic.js",
    },
    "log": false,
    "originalText": "",
    "text": "",
  },
  "previousLocation": Object {
    "column": undefined,
    "line": 3,
    "sourceId": "magic.js",
    "sourceUrl": "http://localhost:8000/magic.js",
  },
}
`;
